---
title: "powsimR: Power Analysis and Sample Size Estimation for Bulk and Single Cell RNA-Seq Experiments"
author:
- name: Beate Vieth
  affiliation: Human Genomics and Anthropology, Department of Biology II, LMU Munich, Germany
  email: vieth.beate@gmail.com
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_float: true
    highlight: pygments
  BiocStyle::pdf_document: default
package: powsimR
bibliography: Bioc.bib
self_contained: no
abstract: |
  This vignette describes the R package `r Githubpkg("bvieth/powsimR")` [@Vieth2017-lh]. It aims to be at once a demonstration of its features and a guide to its usage. Power analysis is essential to optimize the design of RNA-seq experiments and to assess and compare the power to detect differentially expressed genes in RNA-seq data. `r Githubpkg("bvieth/powsimR")` is a flexible tool to simulate and evaluate differential expression from bulk and especially single-cell RNA-seq data making it suitable for a priori and posterior power analyses. For our read count simulations, we (1) reliably model the mean, dispersion and dropout distributions as well as the relationship between those factors from the data. (2) Simulate read counts from the empirical mean-variance and dropout relations, while offering flexible choices of the number of differentially expressed genes, effect sizes, DE testing and normalisation method. (3) Finally, we evaluate the power over various sample sizes and visualise the results with error rates plots.
vignette: |
  %\VignetteIndexEntry{Authoring R Markdown vignettes}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

``` {r knitset, eval=TRUE, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50), 
                      fig.align = 'center', 
                      message=FALSE, error=FALSE, warning=FALSE)
```

**Note:** if you use powsimR in published research, please cite:

> Vieth, B., Ziegenhain, C., Parekh, S., Enard, W. and Hellmann, I. (2017)
> powsimR: Power Analysis for Bulk and Single Cell RNA-Seq Experiments.
> *Bioinformatics*, **33(21)**:3486-88.
> [10.1093/bioinformatics/btx435](http://dx.doi.org/10.1093/bioinformatics/btx435)

# Prerequisites {.unnumbered}

`powsimR` has a number of dependencies that need to be installed before hand (see also the README file on github).

```{r dep, echo = TRUE, eval = FALSE, tidy = FALSE}
ipak <- function(pkg, repository=c('CRAN', 'Bioconductor', 'github')){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  # new.pkg <- pkg
  if (length(new.pkg)) {
    if(repository=='CRAN') {
      install.packages(new.pkg, dependencies = TRUE)
    }
    if(repository=='Bioconductor') {
      if(strsplit(version[['version.string']], ' ')[[1]][3] > "3.5.0"){
        if (!requireNamespace("BiocManager")){
            install.packages("BiocManager")
        }
            BiocManager::install(new.pkg, dependencies=TRUE, ask=FALSE)
      }
      if(strsplit(version[['version.string']], ' ')[[1]][3] < "3.5.0"){
              source("https://bioconductor.org/biocLite.R")
      biocLite(new.pkg, dependencies=TRUE, ask=FALSE)
      }
    }
    if(repository=='github') {
      devtools::install_github(new.pkg, build_vignettes = FALSE, force = FALSE, dependencies=TRUE)
    }
  }
}

# CRAN PACKAGES
cranpackages <- c("broom", "cobs", "cowplot", 
                  "data.table", "doParallel", "dplyr", "DrImpute", 
                  "fastICA", "fitdistrplus", "foreach", "future",
                  "gamlss.dist", "ggplot2", "ggpubr", "grDevices",
                  "grid", "Hmisc", "kernlab", "MASS", "magrittr", "MBESS", "Matrix",
                  "matrixStats", "mclust", "methods", "minpack.lm", "moments", "msir", 
                  "NBPSeq", "nonnest2", "parallel", "penalized", "plyr", "pscl",
                  "reshape2", "Rmagic", "rsvd", "Rtsne", "scales", "Seurat", "snow", "sctransform", 
                  "stats", "tibble", "tidyr", "truncnorm", "VGAM", "ZIM", "zoo")
ipak(cranpackages, repository='CRAN')

# BIOCONDUCTOR
biocpackages <- c("bayNorm", "baySeq", "BiocGenerics", "BiocParallel", 
                  "DEDS", "DESeq2", "EBSeq", "edgeR", "IHW", "iCOBRA",
                  "limma", "Linnorm", "MAST", "monocle", "NOISeq", "qvalue", "ROTS", "RUVSeq", 
                  "S4Vectors", "scater", "scDD", "scde", "scone", "scran", "SCnorm", 
                  "SingleCellExperiment", "SummarizedExperiment", "zinbwave")
ipak(biocpackages, repository='Bioconductor')

# GITHUB
githubpackages <- c('nghiavtr/BPSC', 'cz-ye/DECENT', 
                    'mohuangx/SAVER', 'statOmics/zingeR')
ipak(githubpackages, repository = 'github')
```

After installing the dependencies, `powsimR` can be installed by using devtools as well.

```{r install, echo=TRUE, eval=FALSE, tidy=FALSE}
devtools::install_github('bvieth/powsimR', 
                         build_vignettes = TRUE, 
                         dependencies = FALSE)
library("powsimR")
```


# Introduction

In this vignette, we illustrate the features of `powsimR` by assessing the power to detect differential expression between two groups of mouse embryonic stem cells cultured in standard 2i/LIF culture medium [GSE75790](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75790) [@Ziegenhain2017-sf].

# _powsimR_ Workflow

The basic workflow of `powsimR` is illustrated in Figure \@ref(fig:schematic): A) The mean-dispersion relationship is estimated from RNA-seq data, which can be either single cell or bulk data. The user can provide their own count table or a publicly available one and choose whether to fit a negative binomial or a zero-inflated negative binomial. The plot shows a representative example of the mean-dispersion relation estimated, assuming a negative binomial for the Ziegenhain data, the red line is the loess fit, that we later use for the simulations. B) These distribution parameters are then used to set up the simulations. C) Finally, the error rates are calculated. These can be either returned as marginal estimates per sample configuration (top), or stratified according to the estimates of mean expression, dispersion or dropout rate (bottom). Furthermore, the user can evaluate the analytical choices (e.g. imputation and normalisation).

```{r schematic, fig.cap="PowsimR schematic overview. (A) Estimation (B) Simulation (C) Evaluation.", echo=F, eval=T, include=T, fig.wide = T}
knitr::include_graphics("powsimRvignetteschematic.png")
```

# Parameter Estimation

## Gene Expression

The parameters of the (zero-inflated) negative binomial distribution, i.e. mean and dispersion are estimated by the function `estimateParam`. In addition, the dropout probability, i.e. the fraction of zero counts per gene, is calculated.

It is very important to supply the count matrix of a RNA-seq experiment profiling one group of samples or cells only. For example, for bulk please provide the count matrix containing the expression of the untreated samples / wild type _ONLY_; for single cells please provide the count matrix containing the expression of one cell type sampled from one patient / mouse _ONLY_, e.g. the expression of monocytes from peripheral mononuclear blood (PBMC) extracted from one patient.

The user can choose between two estimation frameworks:

* Negative binomial distribution (NB)
* Zero-inflated negative binomial distribution (ZINB)

In both cases matching moments estimation of mean and dispersion are based on normalized counts.

The user can choose between multiple normalisation methods (see Details section of `estimateParam`). Furthermore, a number of methods are group sensitive (e.g. batch labels can be provided in SCnorm). For single cell data, it is important to filter out genes and/or cells that are outliers or deemed undetected (see options SampleFilter and GeneFilter).

The estimates, sequencing depth and normalisation factors are plotted with `plotParam`.

With the following command, we estimate and plot the parameters for the mouse embryonic stem cells profiled with Smartseq2 [@Ziegenhain2017-sf] (Figure \@ref(fig:paramsplot)). As expected for single cell RNA-seq, the variability (i.e. dispersion) and dropout rates are high. Furthermore, the dispersion strongly depends on the mean and does not level off with higher mean values compared to bulk data.

```{r params, echo=T, eval=T, include=T}
library("powsimR")
data("SmartSeq2_Gene_Read_Counts")
Batches <- data.frame(Batch = sapply(strsplit(colnames(SmartSeq2_Gene_Read_Counts), "_"), "[[", 1),
                      stringsAsFactors = FALSE, row.names = colnames(SmartSeq2_Gene_Read_Counts))
data("GeneLengths_mm10")
estparam_gene <- estimateParam(countData = SmartSeq2_Gene_Read_Counts,
                          readData = NULL,
                          batchData = Batches,
                          spikeData = NULL,
                          spikeInfo = NULL,
                          Lengths = GeneLengths, MeanFragLengths = NULL,
                          RNAseq = 'singlecell', Protocol = 'Read',
                          Distribution = 'ZINB', Normalisation = "scran",
                          GeneFilter = 0.1, SampleFilter = 3,
                          sigma = 1.96, NCores = NULL, verbose = TRUE)
```

```{r paramsplot, echo=T, eval=T, include=T, fig.height = 15, fig.width=20, fig.cap="Estimated parameters for Ziegenhain data set. A) Quality Control Metrics: Sequencing depth; Library size factors with median (black line) for the filtered data set; Detected genes; Ratio of gene to spike-in counts (if spike-ins were provided). Outliers are marked in red. B) Marginal Distribution of gene mean, dispersion and dropout rate per estimation set. C) Number of genes and samples per estimation set. Provided by the user; Detected = number of genes and samples with at least one count; All = number of genes for which mean, dispersion and dropout could be estimated using non-outlying samples. \nFiltered = number of genes above filter threshold for which mean, dispersion and dropout could be estimated using non-outlying samples. Dropout Genes = number of genes filtered out due to dropout rate. D) Local polynomial regression fit between mean and dispersion estimates with variability band per gene (yellow). Common dispersion estimate (grey dashed line). E) Fraction of dropouts versus estimated mean expression per gene."}
plotParam(estparam_gene, Annot = F)
```

We have implemented a count simulation framework assuming either a negative binomial distribution or a zero-inflated negative binomial distribution.
To predict the dispersion given a random draw of mean expression value observed, we apply a locally weighted polynomial regression fit. To capture the variability of dispersion estimates observed, a local variability prediction band is applied.
For bulk RNA-seq experiments, dropouts are less probable but can still occur. To include this phenomenon we sample from the observed dropout rates for genes that have a mean expression value below 5\% dropout probability determined by a decrease constrained B-splines regresssion of dropout rate against mean expression (`cobs` in `r CRANpkg('cobs')`).
The resulting count matrix has similar distributional characteristics as the original Ziegenhain data set.
For the zero-inflated negative binomial distribution, the mean-dispersion relation is similarly estimated, but based on positive read counts. Furthermore, the dropouts are also predicted based on a locally weighted polynomial regression fit between mean and dropouts. Of note, this fit is done separately for amplified and non-amplified transcripts and similar proportions of genes as observed are also generated in the simulations [@Ziegenhain2017-sf].
We have found that the negative binomial distribution is particularly suited for UMI-methods (e.g. SCRB-Seq, Drop-Seq, 10XGenomics) [@Vieth2017-lh].

## Spike-ins

Some normalisation methods can use spike-ins as part of their normalisation (e.g. SCnorm, scran, Census). To use spike-in information in the simulations, their distributional characteristics need to be estimated. We follow the estimation and simulation framework presented in [@Kim2015-zo] where the variance is decomposed into shot noise and mRNA loss due to capture and sequencing efficiency. In short, the parameters for a Beta-distribution describes the RNA molecule capture efficiency and the parameters of a Gamma distribution describes the sequencing efficiency, which we can then use to simulate in silico spike-ins given a mean expression value. We assume that biological variance does not contribute to spike-in expression.
The user needs to define the spike-in expression table and the spike-in information table (IDs, molecules, lengths per spike-in) in the function `estimateSpike`.
The following formula can help the user to calculate the number of molecules of spike-ins:

\begin{equation}
    Y_{j} = c_{j} * V * 10^{-3} * D^{-1} * 10^{-18} * {Avogadro}, \quad j=1,\dots,92
\end{equation}

The number of molecules $Y_{j}$ for each ERCC spike-in species is the product of the molar concentration $c_{j}$ (attomoles per microlitre), the dilution factor $1/D$, the volume $V$ (nanolitre), Avogadros' constant ($6.02214129*10^{23}$) and conversion factors between unit scales.

With the following command, we estimate the parameters for the spike-ins added to the library preparation of mouse embryonic stem cells cultured in standard 2i+lif medium [@Ziegenhain2017-sf]. Descriptive plots of the spike-ins can be drawn with `plotSpike`  (Figure \@ref(fig:spikeplot)).

```{r spikeparams, echo=T, eval=T, include=T, warning=F}
data("SmartSeq2_SpikeIns_Read_Counts")
data("SmartSeq2_SpikeInfo")
Batches = data.frame(Batch = sapply(strsplit(colnames(SmartSeq2_SpikeIns_Read_Counts), "_"), "[[", 1),
                       stringsAsFactors = F,
                       row.names = colnames(SmartSeq2_SpikeIns_Read_Counts))
# estimation
estparam_spike <- estimateSpike(spikeData = SmartSeq2_SpikeIns_Read_Counts,
spikeInfo = SmartSeq2_SpikeInfo,
MeanFragLength = NULL,
batchData = Batches,
Normalisation = 'depth')

```

```{r spikeplot, fig.cap="Estimated parameters for the spike-ins added to Smartseq2 libraries in Ziegenhain dataset. (A) Sequencing depth per sample with median sequencing depth (grey dashed line). (B) Library size normalisation factor per sample with median size factor (grey dashed line). (C) Calibration curve with mean expression estimates and average R squared over all cells. (D) Capture efficiency with binomial logistic regression fit over all cells.", echo=T, warning=F, eval=T, include=T, fig.height = 7, fig.width=10, fig.align='centered'}
plotSpike(estSpike = estparam_spike, Annot = FALSE)
```

# Simulations

For simulating differential expression between two groups, the number of genes, number of simulations, percentage of differential expression and effect size are set up with the function `Setup`. The effect size is here defined as the log2 fold change which can be a constant, sampled from a vector or function. The uniform, normal and gamma distributions are possible options and illustrated in figure \@ref(fig:lfcs). Depending on the settings, these distribution can be broader or narrower. If using this option, we recommend to choose a distribution that closely resembles previously observed or expected fold changes. 
There is also the possibility to include batch effects.

```{r lfcs, fig.cap="Log2 fold change examples for gamma, uniform and normal distribution.", echo=F, eval=F, include=T, out.width = "95%"}
knitr::include_graphics("lfcdist.jpeg")
```

The number of replicates per group (n1 and n2) are also defined in `Setup` which can be unbalanced. In addition the user can choose to include the simulaiton of dropout genes so that the resulting count matrix as a percentage of dropouts equal to the rate filtered out during estimation.

The distribution estimates and these settings are then combined to one object.
The following command sets up simulations with 10,000 genes, 20\% genes being DE, log fold change sample from a narrow gamma distribution and parameter estimates based on Smartseq2 libraries in Ziegenhain dataset. Please note that I decreased the number of simulations for this illustration.

```{r simsetup, echo = TRUE, eval = TRUE}
# define log fold change
p.lfc <- function(x) sample(c(-1,1), size=x,replace=T)*rgamma(x, shape = 1, rate = 2)
# set up simulations
setupres <- Setup(ngenes = 10000, nsims = 10,
                  p.DE = 0.1, pLFC = p.lfc,
                  n1 = c(48, 96, 384, 800), n2 = c(48, 96, 384, 800),
                  Thinning = NULL, LibSize = 'given',
                  estParamRes = estparam_gene,
                  estSpikeRes = estparam_spike,
                  DropGenes = TRUE,
                  sim.seed = 5299, verbose = TRUE)

```

With the setup defined, the differential expression simulation is run with `simulateDE`. For this, the user needs to set the following options:

* **DEmethod:** The differential testing method. The user can choose between 20 methods in total.  10 developed for bulk, 9 developed for single cells (see the Details section of `simulateDE`).
* **Normalisation:** The normalisation method. The user can choose between 10 methods in total.  5 developed for bulk, 5 developed for single cells (see the Details section of `simulateDE`).

There are also additional options: Whether to apply a prefiltering or imputation step prior to normalisation; whether spike-in information should be used (if available). For more information, please consult the Details section of `simulateDE`.

``` {r simrun, eval=T, echo=T}
simres <- simulateDE(SetupRes = setupres,
                     Prefilter = NULL, Imputation = NULL,
                     Normalisation = 'scran',
                     DEmethod = "limma-trend", DEFilter = FALSE,
                     NCores = NULL, verbose = TRUE)
```

# Evaluation

The results of differential expression simulation are evaluated with `evaluateDE`. We have separated the evaluation from DE detection to allow the user to evaluate power in a comprehensive way as advocated by [@Wu2015-uj].
In this function, the proportions and error rates are estimated. The rates can be stratified by mean, dispersion, dropout or log fold change. Furthermore, the user can choose between different multiple testing correction methods (see `p.adjust.methods`, `ihw` in `r Biocpkg("IHW")` and `qvalue` in `r Biocpkg("qvalue")`). Also, the genes can be filtered by mean, dispersion or dropout. To define biologically interesting genes, a cutoff for the log fold change with `delta` can be set.

With the following command we evaluate the marginal TPR and FDR conditional on the mean expression for the simulation based on Ziegenhain data set. 

```{r, echo = T, eval=F}
evalres = evaluateDE(simRes = simres,
                     alpha.type = 'adjusted',
                     MTC = 'BH',
                     alpha.nominal = 0.1,
                     stratify.by = 'mean',
                     filter.by = 'none',
                     strata.filtered = 1,
                     target.by = 'lfc',
                     delta = 0)
```

The results of the evaluation can be plotted with `plotEvalDE`.

* **rate:** The user can choose to plot the marginal or stratified error rates. The number of genes per stratum are also summarised.
* **quick:** If this is set to `TRUE` then only the TPR and FDR will be plotted.

With the following commands, the quick marginal and conditional power assessment for the Ziegenhain data is plotted (Figurer \@ref(fig:evalplot1) and \@ref(fig:evalplot2)).

```{r evalplot1, echo=T, eval=T, fig.cap="Marginal Error Rates. (A) Marginal TPR and FDR per sample size comparison. (B) Marginal TPR and FDR per sample size comparison with dashed line indicating nominal alpha level (type I error) and nominal 1-beta level, i.e. 80% power (type II error)."}
plotEvalDE(evalRes = evalres,
            rate='marginal',
            quick=TRUE, Annot=FALSE)
```

```{r evalplot2, echo=T, eval=T, fig.cap="Stratified Error Rates. (A) Conditional TPR and FDR per sample size comparison per stratum. (B) Number of equally (EE) and differentially expressed (DE) genes per stratum."}
plotEvalDE(evalRes = evalres,
            rate='stratified',
            quick=TRUE, Annot=FALSE)
```

# Additional Functionalities

## Simulation Settings

### Include differences in library sizes

By default, there is no difference in library sizes between the samples. If the user wishes for a more realistic, i.e. more variable distribution of read counts across samples, the library sizes can be sampled from observed, vector or function. See the Details section of `Setup`.

### Downsample simulated count matrices by binomial thinning

There is the additional option to downsample simulated count matrices with the option `Thinning` in `Setup` in order to know in how far a shallower sequencing depth affects the power to detect differential expression. For example, the estimation was done using 500 cells and in the simulations we would like to compare two groups with a total of 1000 cells but do not want to increase the sequencing depth compared to the pilot data. Thus we can set the Thinning parameters to 0.5 to half the number of counts per cell. Please note, that the thinning assumes that the count matrix is based on reads, not UMIs. For UMI-methods, please supply the UMI count matrix as `countData` in `estimateParam` and the raw read count matrix without UMI deduplication as `readData`. For more information, please check out the help page of `estimateParam`.

## Negative Binomial Parameters

### Count matrices of single cell RNA-seq experiments

We have uploaded count matrices of 5 single cell RNA-seq experiments on [github](https://github.com/bvieth/powsimRData). There are also smaller data sets included, see description on the help page `GeneCounts`.
The user can calculate the negative binomial parameters with `estimateParam`, view these estimates with `plotParam` and use it as an input for simulations.

### Access to raw read counts stored in online data base

We have provided a number of exemplatory single cell RNA-seq data sets for parameter estimation. Nevertheless, you might not find a data set that matches your own experimental setup. In those cases, we recommend to check online repositories for a suitable data set. Below you can find an example script to get count tables from [recount2](https://jhubiostatistics.shinyapps.io/recount/) [@Collado-Torres2017-mo]. For a single cell RNA-seq data base, see [conquer](http://imlspenticton.uzh.ch:3838/conquer/) [@Soneson2018-xe].
As before, the user can then estimate the negative binomial parameters with `estimateParam`, view these estimates with `plotParam` and use it as an input for `Setup`.

```{r online_repos, echo = T, eval = F}
# Install and load the R package
BiocManager::install("recount")
library('recount')

# Download the data set
url <- download_study('SRP060416')

# Load the data
load(file.path('SRP060416', 'rse_gene.Rdata'))

# count table
cnts <- assay(rse_gene)
# sample annotation
sample.info <- data.frame(colData(rse_gene)@listData, stringsAsFactors=F)
# gene annotation
gene.info <- data.frame(GeneID=rowData(rse_gene)@listData$gene_id, GeneLength=rowData(rse_gene)@listData$bp_length, stringsAsFactors=F)
```

## Evaluate distributions for RNA-seq data sets

It is important to validate the appropiateness of the chosen simulation framework. The function `evaluateDist` compares the theoretical fit of the Poisson, negative binomial, zero-inflated Poisson and zero-inflated negative binomial and beta-Poisson distribution to the empirical RNA-seq read counts ([@Colin_Cameron2013-vb], [@Kim2013-qo], [@Delmans2016-ef]).
The evaluation is then plotted with the function `plotEvalDist` which summarizes the best fitting distribution per gene based on goodness-of-fit statistics (Chi-square test), Akaike Information Criterium, comparing observed dropouts with zero count prediction of the models and comparing the model fitness with Likelihood Ratio Test and Vuong Test.
As noted by other developers, goodness-of-fit tests are not an objective tool and heavily depend on sample sizes [@Delignette-Muller2015-ie]. A graphical evaluation of the fitted distribution is considered the most appropiate way but for high-throughput sequencing this is an unrealistic recommendation.
Bulk RNA-seq experiments are usually conducted with a small number of samples. We therefore recommend to rely on the goodness-of-fit validation by [@Mi2015-ri]. 

With the following command, we determine and plot the fitting for the embryonic stem cells cultured in standard 2i+lif medium profiled with the Smartseq2 protocol [@Ziegenhain2017-sf]. Note that the results shown in Figure \@ref(fig:evaldistplot) will differ from the executed command since I reduced the data set and evaluate the fitting only for a fraction of genes.

``` {r evaldist, eval=F, echo=T}
data("SmartSeq2_Gene_Read_Counts")
evalDistRes <- evaluateDist(countData = SmartSeq2_Gene_Read_Counts, 
                            batchData = NULL,
                            spikeData = NULL, spikeInfo = NULL,
                            Lengths = NULL, MeanFragLengths = NULL,
                            RNAseq = "singlecell", Protocol = "UMI",
                            Normalisation = "scran",
                            GeneFilter = 0.1, SampleFilter = 3,
                            FracGenes = 0.1,
                            verbose = TRUE)
plotEvalDist(evalDistRes)
```

```{r evaldistplot, echo=F, eval=F, include=T, fig.wide = TRUE, fig.cap="Distribution Evaluation. A) Goodness-of-fit of the model assessed with a Chi-Square Test based on residual deviance and degrees of freedom. B) Akaike Information Criterion per gene: Model with the lowest AIC. Model with the lowest AIC and passed goodness-of-fit statistic test.  C) Observed versus predicted dropouts per model and gene plotted without outliers. D) Model Assessment based on LRT for nested models and Vuong test for nonnested models. "}
knitr::include_graphics("evaldist_smartseq2.png", error = FALSE)
```

# Session info {.unnumbered}

Here is the output of `sessionInfo()` on the system on which this
document was compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References {.unnumbered}
